### PR TITLE
Audio DSP power gating and clock gating enable

### PR DESCRIPTION
**Pull request by ranj063**
Enable clock gating and power gating after fw boot

### CODE REVIEW

File: sound/soc/sof/intel/hda.h [hunk 1]
```diff
@@ -17,9 +17,13 @@
 #define PCI_TCSEL			0x44
 #define PCI_CGCTL			0x48
```
Comment (keyonjie) [inline sound/soc/sof/intel/hda.h:19]: better to correct naming of these 2 registers:
```
#define PCI_PGCTL			0x44
#define PCI_CGCTL			0x48
```

File: sound/soc/sof/intel/hda.h [hunk 2]
```diff
@@ -17,9 +17,13 @@
...
 
+/* PCI_TVSEL bits */
+#define PCI_TCSEL_ADSPPGD               BIT(2)
+
 /* PCI_CGCTL bits */
 #define PCI_CGCTL_MISCBDCGE_MASK	BIT(6)
 #define PCI_CGCTL_LSRMD_MASK		BIT(4)
```
Comment (keyonjie) [inline sound/soc/sof/intel/hda.h:25]: this bit should be PGCTL one, rename and move it to group /* PCI_PGCTL bits */

File: sound/soc/sof/intel/hda.h [hunk 3]
```diff
@@ -17,9 +17,13 @@
 #define PCI_TCSEL			0x44
 #define PCI_CGCTL			0x48
 
+/* PCI_TVSEL bits */
+#define PCI_TCSEL_ADSPPGD               BIT(2)
+
```
Comment (keyonjie) [inline sound/soc/sof/intel/hda.h:22]: rename this to PCI_PGCTL_ADSPPGD

File: sound/soc/sof/intel/hda-ctrl.c [hunk 4]
```diff
@@ -148,6 +149,24 @@ void hda_dsp_ctrl_misc_clock_gating(struct snd_sof_dev *sdev, bool enable)
 	snd_sof_pci_update_bits(sdev, PCI_CGCTL, PCI_CGCTL_MISCBDCGE_MASK, val);
 }
 
+void hda_dsp_ctrl_clock_power_gating(struct snd_sof_dev *sdev, bool enable)
+{
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	u32 val;
+
+	/* Update PDCGE bit of CGCTL register */
+	val = enable ? PCI_CGCTL_ADSPDCGE : 0;
```
Comment (keyonjie) [inline sound/soc/sof/intel/hda-ctrl.c:158]: I think this is opposite for PCI_TCSEL_ADSPPGD?  

Be careful that some bits are gating disable(e.g. ADSPPGD), some others are gating enable(e.g. PCI_CGCTL_ADSPDCGE), we need to handle them carefully.

File: sound/soc/sof/sof-priv.h [hunk 5]
```diff
@@ -83,6 +83,7 @@ struct snd_sof_dsp_ops {
 	int (*resume)(struct snd_sof_dev *sof_dev);
 	int (*runtime_suspend)(struct snd_sof_dev *sof_dev, int state);
 	int (*runtime_resume)(struct snd_sof_dev *sof_dev);
+	void (*clock_power_gating)(struct snd_sof_dev *sof_dev, bool enable);
```
Comment (lgirdwood) [inline sound/soc/sof/sof-priv.h:86]: We should name this a little more generic so we can handle more than just clocks. It does look like here we are entering an idle state ? If so snd_sof_dsp_idle(sdev, bool idle) . Please also don't name the variable `enable` here, best to use the verb name.

Comment (ranj063) [reply sound/soc/sof/sof-priv.h:86]: @lgirdwood we are enabling clock gating and power for the audio DSP here i.e. the HW can power gate when it is idle. So calling it idle would be misleading as typically in the context of power management, idle would indicate we are decrementing the ref count.

File: sound/soc/sof/intel/hda-ctrl.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-ctrl.c:158]: @keyonjie yes, but the code does the right thing here. When we want to enable clock and power gating we write 1 to PCI_CGCTL_ADSPDCGE in CGCTL and write 0 to PCI_PGCTL BIT(2).

File: sound/soc/sof/intel/hda.h (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda.h:25]: sure, that will avoid confusion!

File: sound/soc/sof/sof-priv.h (hunk 5 above)
Comment (lgirdwood) [reply sound/soc/sof/sof-priv.h:86]: @ranj063 is the DSP HW in D3 state here when we gate the clocks ?If so, shouldn't this be part of suspend/resume calls at the HW abstraction level.

Comment (plbossart) [reply sound/soc/sof/sof-priv.h:86]: @ranj063 I am also not clear on the DSP state. If you do power gating why do you need clock gating? Isn't it more when you are in D0 (or D0ix) that you allow the clocks to be gated to allow parts of the hardware to save power?

Comment (ranj063) [reply sound/soc/sof/sof-priv.h:86]: > @ranj063 is the DSP HW in D3 state here when we gate the clocks ?If so, shouldn't this be part of suspend/resume calls at the HW abstraction level.

@lgirdwood the HW is not in D3 here. And we not performing power gating or clock gating here either. We are just enabling them meaning that we are indicating to the HW that if the audio dsp is idle, it should be clocked gated and power gated. When the audio device is idle and in D3, these bits will enable the HW to power gate/clock gate. 

Before we boot the firmware, we disable power gating during dsp probe. So if we dont re-enable it, I think we will never be able to power gate the audio dsp.

Comment (ranj063) [reply sound/soc/sof/sof-priv.h:86]: > @ranj063 I am also not clear on the DSP state. If you do power gating why do you need clock gating? Isn't it more when you are in D0 (or D0ix) that you allow the clocks to be gated to allow parts of the hardware to save power?

@plbossart we are just enabling the HW to be able to do clock gating and power gating here. The dsp is not in D3 after these bits are set. But when an opportunity arises the HW will do the needful if these bits are set.

Comment (ranj063) [reply sound/soc/sof/sof-priv.h:86]: @lgirdwood @plbossart btw i didnt invent anything new here. I just copied it from the skylake driver. We seem to have missed this sequence earlier.

Comment (lgirdwood) [reply sound/soc/sof/sof-priv.h:86]: @ranj063 ok, so setting these bits allows the HW to opportunistically gate clocks/PM resources when the FW is idle after boot is complete ? If so, should probably be part of the existing HW abstraction ops for fw ready and boot ?

Comment (ranj063) [reply sound/soc/sof/sof-priv.h:86]: @lgirdwood thats correct. I think it makes sense to make it part of the "run" ops. Let me make the change.

**Comment by ranj063** [comment]
closing this PR now. Will open a new one with more comprehensive PM updates.

**Comment by zhuyingjiang** [comment]
link to #249

**Pull request closed without merging**